library/core/src/iter/adapters/step_by.rs

```diff
@@ -233,3 +225,96 @@ where
+
+trait SpecRange: Iterator {
+    fn spec_fold<Acc, F>(self, acc: Acc, f: F) -> Acc
+    where
+        F: FnMut(Acc, Self::Item) -> Acc;
+}
+
+impl<I> SpecRange for StepBy<I>
+where
+    I: Iterator,
+{
+    default fn spec_fold<Acc, F>(mut self, mut acc: Acc, mut f: F) -> Acc
+    where
+        F: FnMut(Acc, Self::Item) -> Acc,
+    {
+        #[inline]
+        fn nth<I: Iterator>(iter: &mut I, step: usize) -> impl FnMut() -> Option<I::Item> + '_ {
+            move || iter.nth(step)
+        }
+
+        if self.first_take {
+            self.first_take = false;
+            match self.iter.next() {
+                None => return acc,
+                Some(x) => acc = f(acc, x),
+            }
+        }
+        from_fn(nth(&mut self.iter, self.step)).fold(acc, f)
+    }
+}
+
+macro_rules! spec_int_ranges {
+    ($($t:ty)*) => ($(
+        impl SpecRange for StepBy<Range<$t>> {
+            #[inline]
+            fn spec_fold<Acc, F>(self, mut acc: Acc, mut f: F) -> Acc
+            where
+                F: FnMut(Acc, Self::Item) -> Acc
+            {
+                use crate::convert::TryFrom;
+
+                let mut range = self.iter;
+
+                if !self.first_take {
+                    // `next()` implements `increment-by-step, get-and-increment-by-one`, leaving
+                    // range.start set to one element after the previously returned one.
+                    // This fold uses `get, increment-by-step+1` which leaves the value
+                    // set to one for the next iteration.
+                    //
+                    // Here we get back into the desired state if the iterator
+                    // has been partially advanced
+
+                    // Don't care about overshooting the range end since it'll be dropped at the end
+                    // of the function
+                    range.start = range.start.saturating_add(<$t>::try_from(self.step).unwrap_or(<$t>::MAX));
+                }
+
+                // if a step size larger than the type has been specified fall back to
+                // t::MAX which will never yield more than 1 item.
+                let step = <$t>::try_from(self.step + 1).unwrap_or(<$t>::MAX);
+                // if T is narrower or equal to usize then Range<T> can never have a length > T::MAX
+                let inner_len = <$t>::try_from(range.size_hint().0).unwrap_or(<$t>::MAX);
+                let yield_count = inner_len.div_ceil(step);
+
+                if yield_count == 0 {
+                    return acc;
+                }
+
+                let end = range.start + (yield_count - 1) * step;
+                let mut i = range.start;
+
+                loop {
+                    acc = f(acc, i);
+                    if i == end {
+                        break;
+                    }
+                    i += step;
+                }
+
+                acc
+            }
+        }
+    )*)
+}
+
+#[cfg(target_pointer_width = "64")]
```

Comment:
Pondering: if the specialization is only ever used when the range fits in `usize`, might it make sense to convert to `usize` always, do the looping in that -- it's presumably a good type for the machine anyway -- and just cast to smaller when calling `f`?

Reply:
I'm having some bistability issues with the benchmarks, but at least when targeting x86-64-v3 then using `$t`  different assembly compared to `usize`.


u16 as u16, this also results in better walltime results:

```
       │       vmovd        %ecx,%xmm0                                                                                                                                                                                                                                                                                      ▒
  0.49 │       vpbroadcastw %xmm0,%ymm0                                                                                                                                                                                                                                                                                     ▒
       │       vpaddw       -0x16a5e1(%rip),%ymm0,%ymm0        # e9a0 <str.1-0x660>                                                                                                                                                                                                                                         ▒
       │       mov          %r8d,%esi                                                                                                                                                                                                                                                                                       ▒
       │       shl          $0x6,%esi                                                                                                                                                                                                                                                                                       ▒
       │       lea          (%rcx,%rsi,1),%r10d                                                                                                                                                                                                                                                                             ▒
       │       vxorpd       %xmm1,%xmm1,%xmm1                                                                                                                                                                                                                                                                               ▒
       │       mov          %r8d,%r9d                                                                                                                                                                                                                                                                                       ▒
  0.97 │       vpxor        %xmm2,%xmm2,%xmm2                                                                                                                                                                                                                                                                               ▒
       │       vpxor        %xmm3,%xmm3,%xmm3                                                                                                                                                                                                                                                                               ▒
       │       vpxor        %xmm4,%xmm4,%xmm4                                                                                                                                                                                                                                                                               ▒
       │       vmovdqa      str.1-0xae0,%ymm10                                                                                                                                                                                                                                                                              ▒
       │       vmovdqa      str.1-0x15e0,%ymm7                                                                                                                                                                                                                                                                              ▒
       │       vmovdqa      str.1-0x400,%ymm11                                                                                                                                                                                                                                                                              ▒
       │       cs           nopw 0x0(%rax,%rax,1)                                                                                                                                                                                                                                                                           ▒
  4.90 │b40:┌─→vpaddw       %ymm0,%ymm14,%ymm5                                                                                                                                                                                                                                                                              ▒
  0.97 │    │  vpaddw       %ymm0,%ymm15,%ymm6                                                                                                                                                                                                                                                                              ▒
       │    │  vpxor        %ymm5,%ymm6,%ymm5                                                                                                                                                                                                                                                                               ▒
 18.53 │    │  vpaddw       %ymm0,%ymm12,%ymm6                                                                                                                                                                                                                                                                              ▒
  0.49 │    │  vpaddw       %ymm2,%ymm5,%ymm2                                                                                                                                                                                                                                                                               ▒
  0.49 │    │  vpaddw       %ymm0,%ymm10,%ymm5                                                                                                                                                                                                                                                                              ▒
       │    │  vpxor        %ymm6,%ymm5,%ymm5                                                                                                                                                                                                                                                                               ▒
  9.27 │    │  vpaddw       %ymm0,%ymm13,%ymm6                                                                                                                                                                                                                                                                              ▒
  2.93 │    │  vpaddw       %ymm3,%ymm5,%ymm3                                                                                                                                                                                                                                                                               ▒
  1.95 │    │  vpaddw       %ymm7,%ymm0,%ymm5                                                                                                                                                                                                                                                                               ▒
  0.97 │    │  vpxor        %ymm6,%ymm5,%ymm5                                                                                                                                                                                                                                                                               ▒
 11.18 │    │  vpcmpeqd     %ymm6,%ymm6,%ymm6                                                                                                                                                                                                                                                                               ▒
       │    │  vpaddw       %ymm4,%ymm5,%ymm4                                                                                                                                                                                                                                                                               ▒
  0.97 │    │  vpaddw       %ymm6,%ymm0,%ymm5                                                                                                                                                                                                                                                                               ▒
  1.94 │    │  vpxor        %ymm0,%ymm5,%ymm5                                                                                                                                                                                                                                                                               ▒
 10.70 │    │  vpaddw       %ymm1,%ymm5,%ymm1                                                                                                                                                                                                                                                                               ▒
  0.97 │    │  vpaddw       %ymm0,%ymm11,%ymm0                                                                                                                                                                                                                                                                              ▒
  0.49 │    │  add          $0xffffffc0,%r9d                                                                                                                                                                                                                                                                                ▒
       │    └──jne          b40                                                                                                                                                                                                                                                                                             ▒
       │       vpaddw       %ymm1,%ymm2,%ymm0                                                                                                                                                                                                                                                                               ▒
       │       vpaddw       %ymm0,%ymm3,%ymm0                                                                                                                                                                                                                                                                               ▒
       │       vpaddw       %ymm0,%ymm4,%ymm0                                                                                                                                                                                                                                                                               ▒
       │       vextracti128 $0x1,%ymm0,%xmm1                                                                                                                                                                                                                                                                                ▒
  2.96 │       vpaddw       %xmm1,%xmm0,%xmm0                                                                                                                                                                                                                                                                               ▒
  1.94 │       vpshufd      $0xee,%xmm0,%xmm1                                                                                                                                                                                                                                                                               ▒
  2.45 │       vpaddw       %xmm1,%xmm0,%xmm0                                                                                                                                                                                                                                                                               ▒
  0.50 │       vpshufd      $0x55,%xmm0,%xmm1                                                                                                                                                                                                                                                                               ▒
  1.46 │       vpaddw       %xmm1,%xmm0,%xmm0                                                                                                                                                                                                                                                                               ▒
       │       vpsrld       $0x10,%xmm0,%xmm1                                                                                                                                                                                                                                                                               ◆
  3.40 │       vpaddw       %xmm1,%xmm0,%xmm0                                                                                                                                                                                                                                                                               ▒
       │       vmovd        %xmm0,%r9d                                                                                                                                                                                                                                                                                      ▒
  0.49 │       cmp          %edi,%r8d                                                                                                                                                                                                                                                                                       ▒
       │     ↑ je           a70                                                                                                                                                                                                                                                                                             ▒
       │       test         $0x38,%dil                                                  
```

u16 as usize:

```
       │       vmovd        %edx,%xmm0                                                                                                                                                                                                                                                                                      ▒
       │       vpbroadcastw %xmm0,%ymm0                                                                                                                                                                                                                                                                                     ▒
       │       vpaddw       %ymm6,%ymm0,%ymm0                                                                                                                                                                                                                                                                               ▒
       │       mov          %r9,%rsi                                                                                                                                                                                                                                                                                        ▒
       │       shl          $0x6,%rsi                                                                                                                                                                                                                                                                                       ▒
  0.73 │       lea          (%rsi,%rdx,1),%r10                                                                                                                                                                                                                                                                              ▒
       │       vxorpd       %xmm1,%xmm1,%xmm1                                                                                                                                                                                                                                                                               ▒
       │       mov          %r9,%r8                                                                                                                                                                                                                                                                                         ▒
       │       nop                                                                                                                                                                                                                                                                                                          ▒
  3.19 │a20:┌─→vpaddw       %ymm7,%ymm0,%ymm2                                                                                                                                                                                                                                                                               ▒
  2.48 │    │  vpxor        %ymm0,%ymm2,%ymm2                                                                                                                                                                                                                                                                               ▒
 51.25 │    │  vpaddw       %ymm1,%ymm2,%ymm1                                                                                                                                                                                                                                                                               ▒
  4.54 │    │  vpaddw       %ymm0,%ymm8,%ymm0                                                                                                                                                                                                                                                                               ▒
  9.94 │    │  add          $0xfffffffffffffff0,%r8                                                                                                                                                                                                                                                                         ▒
       │    └──jne          a20                                                                                                                                                                                                                                                                                             ▒
  0.02 │       vextracti128 $0x1,%ymm1,%xmm0                                                                                                                                                                                                                                                                                ▒
  1.11 │       vpaddw       %xmm0,%xmm1,%xmm0                                                                                                                                                                                                                                                                               ▒
  2.16 │       vpshufd      $0xee,%xmm0,%xmm1                                                                                                                                                                                                                                                                               ▒
  2.26 │       vpaddw       %xmm1,%xmm0,%xmm0                                                                                                                                                                                                                                                                               ▒
  0.12 │       vpshufd      $0x55,%xmm0,%xmm1                                                                                                                                                                                                                                                                               ▒
  1.40 │       vpaddw       %xmm1,%xmm0,%xmm0                                                                                                                                                                                                                                                                               ▒
       │       vpsrld       $0x10,%xmm0,%xmm1                                                                                                                                                                                                                                                                               ▒
  1.55 │       vpaddw       %xmm1,%xmm0,%xmm0                                                                                                                                                                                                                                                                               ▒
       │       vmovd        %xmm0,%r8d                                                                                                                                                                                                                                                                                      ▒
  0.75 │       cmp          %r9,%rdi                                                                                                                                                                                                                                                                                        ▒
       │     ↑ je           900                                                                                                                                                                                                                                                                                             ▒
       │       test         $0x8,%dil                               
```